docs/design/moderation.md

```diff
@@ -0,0 +1,147 @@
+
+### Azure AI Content Safety (Recommended)
+
+Azure AI Content Safety is a comprehensive solution designed to detect harmful user-generated and AI-generated content in applications and services. It is suitable for many scenarios such as online marketplaces, gaming companies, social messaging platforms, enterprise media companies, and K-12 education solution providers.
+
+The Azure AI Content Safety APIs provide comprehensive content moderation capabilities:
+- Text analysis for harmful content (hate speech, sexual content, violence, self-harm)
+- Image analysis for inappropriate visual content
+- Multiple severity levels (0-Safe, 2-Low, 4-Medium, 6-High)
+- Support for blocklists and custom categories
+
+The proposed design below utilizes the AI Content Safety APIs.
+
+## Proposed Architecture
+
+The proposed moderation system would consist of the following components:
+
+1. **Moderation Interface**: A generic `Moderator` interface that defines methods for text and image moderation, allowing for different implementation providers.
+2. **Azure Implementation**: An implementation of the `Moderator` interface that integrates with Azure AI Content Safety APIs.
+3. **Integration Points**: Code that integrates moderation into the message processing flow, both for user input and LLM responses.
+
+### Service Availability Considerations:
+- The proposed design would use a "fail-closed" approach for safety - if the moderation service encounters an error, content would be blocked by default.
+- This ensures potentially problematic content doesn't slip through due to technical issues.
+- However, this means if moderation APIs are unavailable, all LLM interactions would be disabled by default.
+
+**Initial Implementation Approach:**
+- For the first release, we will use the fail-closed approach for simplicity and security.
+
+**Potential Future Considerations (Not in Initial Implementation):**
+- Options for fail-open in trusted environments
+- Graceful degradation mechanisms instead of complete service shutdown
+
+## Proposed Moderation Flow
+
+The proposed implementation would include three primary moderation points:
+
+1. **User Input Moderation**:
+   - When a user sends a message to an AI bot, the message would be checked by the moderation service.
+   - If flagged, the user would receive a notification and the message would not be processed by the LLM.
+   - This happens in `processUserRequestToBot()` in `conversations.go`.
+
+2. **LLM Response Moderation**:
+   - When an LLM generates a response, the complete response would be checked before being displayed to users.
+   - If flagged, users would see a generic message instead of the problematic content.
+   - This happens in `streamResultToPost()` in `post_processing.go`.
+
+3. **Image Moderation**:
```

Comment:
I think in this would be also needed to include moderation in the images generated by the AI. I would expect that eventually we support image generation inside mattermost (now is not supported).